source/common/network/udp_listener_impl.cc

```diff
@@ -0,0 +1,163 @@
+
+#include "event2/listener.h"
+
+namespace Envoy {
+namespace Network {
+
+UdpListenerImpl::UdpListenerImpl(const Event::DispatcherImpl& dispatcher, Socket& socket,
+                                 UdpListenerCallbacks& cb)
+    : BaseListenerImpl(dispatcher, socket), cb_(cb) {
+  event_assign(&raw_event_, &dispatcher.base(), socket.ioHandle().fd(),
+               EV_READ | EV_WRITE | EV_PERSIST, eventCallback, this);
+  event_add(&raw_event_, nullptr);
+
+  if (!Network::Socket::applyOptions(socket.options(), socket,
+                                     envoy::api::v2::core::SocketOption::STATE_BOUND)) {
+    throw CreateListenerException(fmt::format("cannot set post-bound socket option on socket: {}",
+                                              socket.localAddress()->asString()));
+  }
+}
+
+void UdpListenerImpl::disable() { event_del(&raw_event_); }
+
+void UdpListenerImpl::enable() { event_add(&raw_event_, nullptr); }
+
+UdpListenerImpl::ReceiveResult UdpListenerImpl::doRecvFrom(sockaddr_storage& peer_addr,
+                                                           socklen_t& addr_len) {
+  constexpr uint64_t const read_length = 16384;
+
+  Buffer::InstancePtr buffer = std::make_unique<Buffer::OwnedImpl>();
+
+  addr_len = sizeof(sockaddr_storage);
+  memset(&peer_addr, 0, addr_len);
+
+  Buffer::RawSlice slice;
+  const uint64_t num_slices = buffer->reserve(read_length, &slice, 1);
+
+  ASSERT(num_slices == 1);
+  // TODO(conqerAtapple): Use os_syscalls
+  const ssize_t rc = ::recvfrom(socket_.ioHandle().fd(), slice.mem_, read_length, 0,
+                                reinterpret_cast<struct sockaddr*>(&peer_addr), &addr_len);
+  if (rc < 0) {
+    return ReceiveResult{Api::SysCallIntResult{static_cast<int>(rc), errno}, nullptr};
+  }
+
+  slice.len_ = std::min(slice.len_, static_cast<size_t>(rc));
+  buffer->commit(&slice, 1);
+
+  return ReceiveResult{Api::SysCallIntResult{static_cast<int>(rc), 0}, std::move(buffer)};
+}
+
+void UdpListenerImpl::eventCallback(int fd, short flags, void* arg) {
+  RELEASE_ASSERT((flags & (EV_READ | EV_WRITE)),
+                 fmt::format("Unexpected flags for callback: {}", flags));
+
+  UdpListenerImpl* instance = static_cast<UdpListenerImpl*>(arg);
+  ASSERT(instance);
+
+  if (flags & EV_READ) {
+    instance->handleReadCallback(fd);
+  }
+
+  if (flags & EV_WRITE) {
+    instance->handleWriteCallback(fd);
+  }
+}
+
+void UdpListenerImpl::handleReadCallback(int fd) {
+  RELEASE_ASSERT(fd == socket_.ioHandle().fd(),
+                 fmt::format("Invalid socket descriptor received in callback {}", fd));
+
+  sockaddr_storage addr;
+  socklen_t addr_len;
+  ReceiveResult recv_result;
+
+  do {
+    recv_result = doRecvFrom(addr, addr_len);
+    if ((recv_result.result_.rc_ < 0) && (recv_result.result_.rc_ != -EAGAIN)) {
+      cb_.onError(UdpListenerCallbacks::ErrorCode::SyscallError, recv_result.result_.errno_);
+      return;
+    }
+
+  } while (recv_result.result_.rc_ == -EAGAIN);
```

Comment:
I'm confused. Won't this spin? Shouldn't we return if we get EAGAIN and wait for a read event?

Reply:
Good catch.